atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ClusterTopologyValidator.java

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.cassandra.pool.CassandraServer;
+import com.palantir.conjure.java.api.config.service.HumanReadableDuration;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import one.util.streamex.EntryStream;
+import org.apache.thrift.transport.TTransportException;
+import org.immutables.value.Value;
+
+public final class ClusterTopologyValidator {
+    private static final SafeLogger log = SafeLoggerFactory.get(ClusterTopologyValidator.class);
+
+    public ClusterTopologyValidator() {}
+
+    public Set<CassandraServer> getNewHostsWithInconsistentTopologiesAndRetry(
+            Set<CassandraServer> newlyAddedHosts,
+            Map<CassandraServer, CassandraClientPoolingContainer> allHosts,
+            HumanReadableDuration waitTimeBetweenCalls,
+            HumanReadableDuration maxWaitTime) {
+        Retryer<Set<CassandraServer>> retryer = RetryerBuilder.<Set<CassandraServer>>newBuilder()
+                .retryIfResult(servers -> servers.size() == allHosts.size())
+                .retryIfException()
+                .withWaitStrategy(
+                        WaitStrategies.fixedWait(waitTimeBetweenCalls.toMilliseconds(), TimeUnit.MILLISECONDS))
+                .withStopStrategy(StopStrategies.stopAfterDelay(maxWaitTime.toMilliseconds(), TimeUnit.MILLISECONDS))
+                .build();
+        try {
+            return retryer.call(() -> getNewHostsWithInconsistentTopologies(newlyAddedHosts, allHosts));
+        } catch (ExecutionException | RetryException e) {
+            throw new SafeRuntimeException("Failed to obtain consistent view of hosts from cluster.", e);
+        }
+    }
+
+    public Set<CassandraServer> getNewHostsWithInconsistentTopologies(
```

Comment:
Cases to consider:
1. All Cassandra hosts are online, and support the endpoint. 
2. Cassandra is performing a rolling restart, and support the endpoint.
3. Cassandra is offline, and support the endpoint.
4. All Cassandra hosts are online, but does not support the new endpoint.
5. Cassandra is performing a rolling restart, but does not support the new endpoint.
6. All Cassandra hosts are online, but only a portion support the new endpoint. 
7. Cassandra is performing a rolling restart, but only a portion support the new endpoint. 

Cases 1-7 need to be handled for: client first-time startup and when a new node joins the cluster. 

** Rolling restart = could be 1 node, or could 1/3 of nodes.